File: java/src/main/java/problems101_200/BinaryTreeZigzagLevelOrderTraversal.java
package problems101_200;

import concepts.TreeNode;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

enum BinaryTreeZigzagLevelOrderTraversal {
    DEFAULT {
        @Override
        List<List<Integer>> zigzagLevelOrder(TreeNode root) {
            List<List<Integer>> lists = new ArrayList<>();
            if (root == null) {
                return lists;
            }

            Queue<TreeNode> queue = new LinkedList<>();
            queue.add(root);
            while (!queue.isEmpty()) {
                int size = queue.size();
                List<Integer> level = new ArrayList<>();
                for (int i = 0; i < size; i++) {
                    TreeNode node = queue.remove();
                    level.add(node.val);
                    if (node.left != null) {
                        queue.add(node.left);
                    }
                    if (node.right != null) {
                        queue.add(node.right);
                    }
                }
                lists.add(level);
            }

            for (int i = 0; i < lists.size(); i++) {
                if (i % 2 == 1) {
                    Collections.reverse(lists.get(i));
                }
            }
            return lists;
        }
    };

    abstract List<List<Integer>> zigzagLevelOrder(TreeNode root);
}

File: kotlin/src/main/kotlin/problems101_200/BinaryTreeZigzagLevelOrderTraversal.kt
package problems101_200

import concepts.TreeNode
import java.util.Collections
import java.util.LinkedList

enum class BinaryTreeZigzagLevelOrderTraversal {
    DEFAULT {
        override fun zigzagLevelOrder(root: TreeNode?): List<List<Int>> {
            val lists = arrayListOf<List<Int>>()
            if (root == null) {
                return lists
            }

            val queue = LinkedList<TreeNode?>()
            queue += root
            while (!queue.isEmpty()) {
                val size = queue.size
                val level = arrayListOf<Int>()
                for (i in 0 until size) {
                    val node = queue.remove()
                    level += node!!.`val`
                    if (node!!.left != null) {
                        queue += node.left
                    }
                    if (node.right != null) {
                        queue += node.right
                    }
                }
                lists += level
            }

            for (i in lists.indices) {
                if (i % 2 == 1) {
                    Collections.reverse(lists[i])
                }
            }
            return lists
        }
    },
    ;

    abstract fun zigzagLevelOrder(root: TreeNode?): List<List<Int>>
}
